apps/prairielearn/src/lib/editors.js

```diff
@@ -468,105 +480,77 @@ class Editor {
     debug('Checking folders', reverseFolders);
 
     let seenNonemptyFolder = false;
-    async.eachOfSeries(
-      reverseFolders,
-      (folder, index, callback) => {
-        if (!seenNonemptyFolder) {
-          const delPath = path.join(rootDirectory, ...idSplit.slice(0, idSplit.length - 1 - index));
-          debug('Checking', delPath);
-
-          fs.readdir(delPath, (err, files) => {
-            if (ERR(err, callback)) return;
-
-            // Delete the subfolder if it's empty, otherwise stop here
-            if (files.length > 0) {
-              debug(delPath, 'is nonempty, stopping here.');
-              debug('Folder contains', files);
-              seenNonemptyFolder = true;
-              callback(null);
-            } else {
-              debug('No files, deleting', delPath);
-              fs.remove(delPath, (err) => {
-                if (ERR(err, callback)) return;
-                callback(null);
-              });
-            }
-          });
+    for (const [index] of reverseFolders.entries()) {
+      if (!seenNonemptyFolder) {
+        const delPath = path.join(rootDirectory, ...idSplit.slice(0, idSplit.length - 1 - index));
+        debug('Checking', delPath);
+
+        const files = await fs.readdir(delPath);
+
+        // Delete the subfolder if it's empty, otherwise stop here
+        if (files.length > 0) {
+          debug(delPath, 'is nonempty, stopping here.');
+          debug('Folder contains', files);
+          seenNonemptyFolder = true;
         } else {
-          callback(null);
+          debug('No files, deleting', delPath);
+          await fs.remove(delPath);
         }
-      },
-      (err) => {
-        if (ERR(err, callback)) return;
-        callback(null);
       }
-    );
+    }
   }
 
   /**
    * Get all existing shortnames, recursing on nonempty directories that do not contain
    * an ".info" file.
    * @param rootDirectory Directory to start searching from.
    * @param infoFile Name of the info file, will stop recursing once a directory contains this.
-   * @param callback {function(err, files)} Function that is called once this has completed.
    */
-  getExistingShortNames(rootDirectory, infoFile, callback) {
+  async getExistingShortNames(rootDirectory, infoFile) {
     let files = [];
-    const walk = (relativeDir, callback) => {
-      fs.readdir(path.join(rootDirectory, relativeDir), (err, directories) => {
-        if (err && (err.code === 'ENOENT' || err.code === 'ENOTDIR')) {
-          // If the directory doesn't exist, then we have nothing to load
-          return callback(null);
+    const walk = async (relativeDir) => {
+      let directories;
+      try {
+        directories = await fs.readdir(path.join(rootDirectory, relativeDir));
+      } catch (err) {
+        // If the directory doesn't exist, then we have nothing to load
+        if (err.code === 'ENOENT' || err.code === 'ENOTDIR') {
+          return;
+        }
+        throw err;
```

Comment:
Aside: this is basically the only case where I don't like async/await. I've written a little function like this in the past to make this slightly less annoying:

```ts
async function wrapErrorfn) {
  try {
    const data = await fn;
    return [data, null];
  } catch (error) {
    return [null, error];
  }
}
```

So you can then do something like this:

```ts
const [directories, err] = await wrapError(fs.readdir());
```

This is roughly what golang does, and not that far off from something like [Rust's `Result` type](https://doc.rust-lang.org/std/result/). We can also add TS generic types so that this works generically, informs `tsc` that the first value is not null if the second is, etc. Do you completely hate this? 😅

Reply:
I was reading about Promise.allSettled the other day, which always resolves with an object that can be used in something similar to what you have in mind. Does that help? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled

Reply:
Interesting, I hadn't thought about applying that function that way. It's actually not bad:

```ts
const [res] = await Promise.allSettled([fs.readdir(path.join(rootDirectory, relativeDir))]);
if (res.status === "rejected") {
  // If the directory doesn't exist, then we have nothing to load
  if (res.reason.code === "ENOENT" || res.reason.code === "ENOTDIR") {
    return;
  }
  throw res.reason;
}
const directories = res.value;
```

Wrapping + unwrapping with arrays is a little annoying; a hypothetical `Promise.settled()` would be nice, and would probably end up looking a lot like what I proposed above.

Compared to the proposed `wrapError`-like thing:

```ts
const [directories, err] = await wrapError(fs.readdir(path.join(rootDirectory, relativeDir)));
if (err) {
  // If the directory doesn't exist, then we have nothing to load
  if (err.code === "ENOENT" || err.code === "ENOTDIR") {
    return;
  }
  throw err;
}
```

If we actually called this `settled` and made if behave more like `Promise.allSettled`, we could use it like this:

```ts
const directoriesResult = await settled(fs.readdir(path.join(rootDirectory, relativeDir)));
if (directoriesResult.status === "rejected") {
  // If the directory doesn't exist, then we have nothing to load
  if (directoriesResult.reason.code === "ENOENT" || directoriesResult.reason.code === "ENOTDIR") {
    return;
  }
  throw directoriesResult.reason;
}
const directories = directoriesResult.value;
```

Reply:
What about:

```
      const directories = await fs.readdir(path.join(rootDirectory, relativeDir)).catch((err) => {
        // If the directory doesn't exist, then we have nothing to load
        if (err.code === 'ENOENT' || err.code === 'ENOTDIR') {
          return null;
        }
        throw err;
      });

Reply:
That might actually be the cleanest, I like it!

Reply:
Done in 8254a4e.